SEED-platform/seed
# Add Batch Audit Template Import

#### Any background context you want to provide?
SEED currently has functionality to update an individual property with [Audit Template](https://buildingenergyscore.energy.gov/rp/buildings) building data. The function leverages the Audit Template API endpoint [GET /api/v2/building_sync/download/rp/buildings/:id](https://api.labworks.org/apidoc/v2/building_sync/download.html). The returned XML is used to create a new PropertyState and associate it with the current PropertyView.

#### What's this PR do?
Adds the ability to batch update properties with data from audit template.

From the Data page, a user may now import Audit Template Buildings via the Add Data Files button. A request is sent to the linked Audit Template account to get all buildings associated with the account details at [GET /api/v2/rp/buildings](https://api.labworks.org/apidoc/v2/buildings/index.html). Backend logic determines which audit tempatae buildings match existing properties for the selected cycle. Matching is solely based on the PropertyState field `audit_template_building_id`. The results are displayed to the user where they may sort/filter results and select which properties to continue to update. Selected properties are looped through and XML files are fetched from the Audit Template API endpoint `GET /api/v2/building_sync/download/rp/buildings/:id`. PropertyViews are updated with the corresponding XMLs.

All Audit Template requests are made within the `seed_celery` container. 

The current implementation can lead to very long wait times. For example, a request to get a building xml for a single building takes ~ 5 seconds. Updating 1000 properties would take about an hour and a half to complete.

The only property that exists in both the PropertyState and the Audit Template Tool is `audit_template_building_id` (`id` on Audit Template). This requires SEED properties to be updated with correct `audit_tempalte_building_id`s which can be time consuming.

If the matching criteria were to change, the audit template get_buildings response must be updated to include properties like `pm_property_id`, `ubid`, or `custom_id`. Below is the current response structure.

```
GET /api/v2/rp/buildings
[{
    "id": 369,
    "name": "New Buildng",
    "owner": {
        "id": 877,  "email": "[test](mailto:test@nrel.gov)", "first_name": "test",
        "last_name": "test",  "organization": { "id": 9, 
        "name": "National Renewable Energy Laboratory - SEED Project" }, 
        "site_admin": false,  "type": "User" 
     },
    "report": {
      "id": 2, "name": "New York City Energy Efficiency Report",
      "city": { "id": 2, "name": "New York", "state": "NY" }
    },
   "status": "Editing",
   "submissions": [],
   "created_at": "2018-08-22T12:21:43.519-07:00",
   "updated_at": "2023-04-04T12:31:04.100-07:00",
   "html_url": "<html_url>",
   "submission_history_html_url": "<history_url>",
   "xml_url":  "<xml_url>"
  },
   ….
]
```
#### Known Issues 
Filtering the Audit template buildings only supports exact matches at the moment, not '<10', '>20'

#### How should this be manually tested?
Add audit_template_building_ids to existing properties with building ids from your Audit Template Tool account. 
Data > Add Data Files > Import Audit Template Buildings > Confirm > select properties > Update
You should see an informative progress bar and a quick summary of new property views added. Check the property detail of the updated properties to ensure a new building file has been uploaded (`at_{audit_tempalte_building_id}_{YYYYMMDD_Hms}`) and the properties attributes have been updated with Audit Template data

#### What are the relevant tickets?
#3770

#### Screenshots (if appropriate)

![Screenshot 2023-07-26 at 9 41 18 AM](https://github.com/SEED-platform/seed/assets/58446472/fb931de6-a1c1-48a9-aff4-8ead0e2c5923)
![Screenshot 2023-07-26 at 9 41 23 AM](https://github.com/SEED-platform/seed/assets/58446472/1f21b67b-bde2-4ce1-b61e-65da357bec89)
![Screenshot 2023-07-26 at 9 41 34 AM](https://github.com/SEED-platform/seed/assets/58446472/e0927539-e1d0-40ac-94b6-f40533528094)

![Screenshot 2023-07-26 at 9 43 03 AM](https://github.com/SEED-platform/seed/assets/58446472/32ad82ff-3e17-40fc-9260-eee789e80c4d)
![Screenshot 2023-07-26 at 9 43 13 AM](https://github.com/SEED-platform/seed/assets/58446472/b639228d-a40d-4bed-8681-d4a86170d9f2)
![Screenshot 2023-07-26 at 9 43 24 AM](https://github.com/SEED-platform/seed/assets/58446472/888f6165-e95e-4833-a4b1-48633595fb74)
